File: composeApp/src/androidMain/kotlin/com/prekogdevs/fintrack/core/ui/DonutChart.kt
package com.prekogdevs.fintrack.core.ui

import androidx.compose.foundation.Canvas
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.core.graphics.toColorInt
import com.prekogdevs.fintrack.domain.CategorySpend

@Composable
fun DonutChart(
    modifier: Modifier = Modifier,
    categories: List<CategorySpend>
) {
    val total = categories.sumOf { it.amount }.toFloat()
    Canvas(modifier = modifier) {
        val strokeWidth = 40f
        val radius = (size.minDimension - strokeWidth) / 2f
        val center = Offset(size.width / 2f, size.height / 2f)
        var startAngle = -90f

        categories.forEach { spend ->
            val fullSweep = (spend.amount.toFloat() / total) * 360f
            val drawSweep = fullSweep - 2f
            if (drawSweep > 0f) {
                drawArc(
                    color = Color(spend.category.color.toColorInt()),
                    startAngle = startAngle,
                    sweepAngle = drawSweep,
                    useCenter = false,
                    topLeft = Offset(center.x - radius, center.y - radius),
                    size = Size(radius * 2, radius * 2),
                    style = Stroke(width = strokeWidth, cap = StrokeCap.Butt)
                )
            }
            startAngle += fullSweep
        }
    }
}

File: iosApp/iosApp/Core/UI/DonutChart.swift
import SwiftUI
import Shared

struct DonutChart: View {
    let categories: [Shared.CategorySpend]

    var body: some View {
        Canvas { context, size in
            let total = categories.reduce(0.0) { $0 + $1.amount }
            let strokeWidth: CGFloat = 40
            let radius = (min(size.width, size.height) - strokeWidth) / 2
            let center = CGPoint(x: size.width / 2, y: size.height / 2)
            var startAngle = Angle.degrees(-90)

            for spend in categories {
                let fraction = spend.amount / total
                let fullSweep = fraction * 360
                let drawSweep = fullSweep - 2

                if drawSweep > 0 {
                    var path = Path()
                    path.addArc(
                        center: center,
                        radius: radius,
                        startAngle: startAngle,
                        endAngle: startAngle + Angle.degrees(drawSweep),
                        clockwise: false
                    )
                    context.stroke(
                        path,
                        with: .color(spend.category.categoryColor),
                        style: StrokeStyle(lineWidth: strokeWidth, lineCap: .butt)
                    )
                }
                startAngle += Angle.degrees(fullSweep)
            }
        }
    }
}
